### PR TITLE
planner: add the doc for the SEMI_JOIN_REWRITE hint

### PR DESCRIPTION
**Pull request by winoros**
### First-time contributors' checklist 

- [x] I've signed [**Contributor License Agreement**](https://cla-assistant.io/pingcap/docs-cn) that's required for repo owners to accept my contribution.

### What is changed, added or deleted? (Required)

Add the doc about the new hint.

### Which TiDB version(s) do your changes apply to? (Required)

**Tips for choosing the affected version(s):**

By default, **CHOOSE MASTER ONLY** so your changes will be applied to the next TiDB major or minor releases. If your PR involves a product feature behavior change or a compatibility change, **CHOOSE THE AFFECTED RELEASE BRANCH(ES) AND MASTER**.

For details, see [tips for choosing the affected versions (in Chinese)](https://github.com/pingcap/docs-cn/blob/master/CONTRIBUTING.md#版本选择指南).

- [x] master (the latest development version)
- [x] v6.2 (TiDB 6.2 versions)
- [ ] v6.1 (TiDB 6.1 versions)
- [ ] v5.4 (TiDB 5.4 versions)
- [ ] v5.3 (TiDB 5.3 versions)
- [ ] v5.2 (TiDB 5.2 versions)
- [ ] v5.1 (TiDB 5.1 versions)
- [ ] v5.0 (TiDB 5.0 versions)

### What is the related PR or file link(s)?

- This PR is translated from:
- Other reference link(s):

### Do your changes match any of the following descriptions?

- [ ] Delete files
- [ ] Change aliases
- [ ] Need modification after applied to another branch 
- [ ] Might cause conflicts after applied to another branch

### CODE REVIEW

**Review by time-and-fate — APPROVED** [review]

**Review by AilinKid — APPROVED** [review]
LGTM

File: optimizer-hints.md [hunk 1]
```diff
@@ -125,6 +125,42 @@ SELECT /*+ HASH_JOIN(t1, t2) */ * FROM t1，t2 WHERE t1.id = t2.id;
 >
 > `HASH_JOIN` 的别名是 `TIDB_HJ`，在 3.0.x 及之前版本仅支持使用该别名；之后的版本同时支持使用这两种名称，推荐使用 `HASH_JOIN`。
 
+### SEMI_JOIN_REWRITE()
+
+`SEMI_JOIN_REWRITE()` 提示优化器可以将半连接（Semi Join）进行改写，改写为普通的内连接。目前该 Hint 只作用在 EXISTS 子查询。在不进行改写的情况下，SEMI JOIN 在选择 Hash Join 的执行方式时，只能够使用子查询去构建哈希表，因此在子查询比外查询结果集大时，执行速度可能会不及预期。在选择 Index Join 的执行方式时，只能够使用外查询作为驱动表，因此在子查询比外查询结果集小时，执行速度可能会不及预期。在使用了改写之后，便可以扩大选择范围，选择更好的执行方式。
```
Comment (TomShawn) [inline optimizer-hints.md:130]:
```suggestion
`SEMI_JOIN_REWRITE()` 提示优化器将查询语句中的半连接 (Semi Join)  改写为普通的内连接。目前该 Hint 只作用于 `EXISTS` 子查询。

如果不使用该 Hint 进行改写，Semi Join 在选择 Hash Join 的执行方式时，只能够使用子查询构建哈希表，因此在子查询比外查询结果集大时，执行速度可能会不及预期。Semi Join 在选择 Index Join 的执行方式时，只能够使用外查询作为驱动表，因此在子查询比外查询结果集小时，执行速度可能会不及预期。

在使用了 `SEMI_JOIN_REWRITE()` 进行改写后，优化器便可以扩大选择范围，选择更好的执行方式。
```

File: optimizer-hints.md [hunk 2]
```diff
@@ -125,6 +125,42 @@ SELECT /*+ HASH_JOIN(t1, t2) */ * FROM t1，t2 WHERE t1.id = t2.id;
...
+
+{{< copyable "sql" >}}
+
+```sql
+EXPLAIN SELECT * FROM t WHERE EXISTS (SELECT 1 from t1 where t1.a=t.a);
+EXPLAIN SELECT * FROM t WHERE EXISTS (SELECT /*+ SEMI_JOIN_REWRITE() */ 1 from t1 where t1.a=t.a);
```
Comment (TomShawn) [inline optimizer-hints.md:136]:
```suggestion
-- 不使用 SEMI_JOIN_REWRITE() 进行改写
EXPLAIN SELECT * FROM t WHERE EXISTS (SELECT 1 from t1 where t1.a=t.a);

-- 使用 SEMI_JOIN_REWRITE() 进行改写
EXPLAIN SELECT * FROM t WHERE EXISTS (SELECT /*+ SEMI_JOIN_REWRITE() */ 1 from t1 where t1.a=t.a);
```

File: optimizer-hints.md [hunk 3]
```diff
@@ -125,6 +125,42 @@ SELECT /*+ HASH_JOIN(t1, t2) */ * FROM t1，t2 WHERE t1.id = t2.id;
...
+```
+
+```sql
++-----------------------------+---------+-----------+------------------------+---------------------------------------------------+
```
Comment (TomShawn) [inline optimizer-hints.md:140]: 能否把上面两个 EXPLAIN 拆开，分别对应一个执行结果，并说明有没有使用改写呢？这样用户可以看得更清楚。

File: subquery-optimization.md [hunk 4]
```diff
@@ -85,3 +85,5 @@ explain select * from t1 where exists (select * from t2);
 | └─TableFullScan_11     | 10000.00 | cop[tikv] | table:t       | keep order:false, stats:pseudo |
 +------------------------+----------+-----------+---------------+--------------------------------+
 ```
+
+除了上述优化器会自动进行的优化之外，也可以通过 [`SEMI_JOIN_REWRITE`](/optimizer-hints.md#semi_join_rewrite) 进行进一步的改写。在不进行改写的情况下，SEMI JOIN 在选择 Hash Join 的执行方式时，只能够使用子查询去构建哈希表，因此在子查询比外查询结果集大时，执行速度可能会不及预期。在选择 Index Join 的执行方式时，只能够使用外查询作为驱动表，因此在子查询比外查询结果集小时，执行速度可能会不及预期。在使用了改写之后，便可以扩大选择范围，选择更好的执行方式。
```
Comment (TomShawn) [inline subquery-optimization.md:89]: 这里的「上述」是指上面全部的子查询情况吗？
```suggestion
在以上子查询情况下，优化器会自动优化语句执行。除以上情况外，你也可以在语句中添加 [`SEMI_JOIN_REWRITE`](/optimizer-hints.md#semi_join_rewrite) hint 进一步改写语句。

如果不使用 `SEMI_JOIN_REWRITE` 进行改写，Semi Join 在选择 Hash Join 的执行方式时，只能够使用子查询构建哈希表，因此在子查询比外查询结果集大时，执行速度可能会不及预期。Semi Join 在选择 Index Join 的执行方式时，只能够使用外查询作为驱动表，因此在子查询比外查询结果集小时，执行速度可能会不及预期。

使用 `SEMI_JOIN_REWRITE` 改写后，优化器便可以扩大选择范围，选择更好的执行方式。
```

File: optimizer-hints.md (hunk 3 above)
Comment (Oreoxmt) [reply optimizer-hints.md:140]: +1

File: subquery-optimization.md (hunk 4 above)
Comment (winoros) [reply subquery-optimization.md:89]: 啊 这个地方是说的上述优化，不是上述子查询

**Review by TomShawn — APPROVED** [review]
LGTM

File: optimizer-hints.md [hunk 5]
```diff
@@ -125,6 +125,56 @@ SELECT /*+ HASH_JOIN(t1, t2) */ * FROM t1，t2 WHERE t1.id = t2.id;
...
+| └─IndexReader_23(Probe)     | 9990.00 | root      |                        | index:IndexFullScan_22                            |
+|   └─IndexFullScan_22        | 9990.00 | cop[tikv] | table:t, index:idx(a)  | keep order:true, stats:pseudo                     |
++-----------------------------+---------+-----------+------------------------+---------------------------------------------------+
+```
+
+{{< copyable "sql" >}}
+
+```sql
+-- 使用 SEMI_JOIN_REWRITE() 进行改写
+EXPLAIN SELECT * FROM t WHERE EXISTS (SELECT /*+ SEMI_JOIN_REWRITE() */ 1 from t1 where t1.a=t.a);
```
Comment (Oreoxmt) [inline optimizer-hints.md:159]:
```suggestion
EXPLAIN SELECT * FROM t WHERE EXISTS (SELECT /*+ SEMI_JOIN_REWRITE() */ 1 FROM t1 WHERE t1.a = t.a);
```

File: optimizer-hints.md [hunk 6]
```diff
@@ -125,6 +125,56 @@ SELECT /*+ HASH_JOIN(t1, t2) */ * FROM t1，t2 WHERE t1.id = t2.id;
...
+
+如果不使用该 Hint 进行改写，Semi Join 在选择 Hash Join 的执行方式时，只能够使用子查询构建哈希表，因此在子查询比外查询结果集大时，执行速度可能会不及预期。Semi Join 在选择 Index Join 的执行方式时，只能够使用外查询作为驱动表，因此在子查询比外查询结果集小时，执行速度可能会不及预期。
+
+在使用了 `SEMI_JOIN_REWRITE()` 进行改写后，优化器便可以扩大选择范围，选择更好的执行方式。
+
+{{< copyable "sql" >}}
+
+```sql
+-- 不使用 SEMI_JOIN_REWRITE() 进行改写
+EXPLAIN SELECT * FROM t WHERE EXISTS (SELECT 1 from t1 where t1.a=t.a);
```
Comment (Oreoxmt) [inline optimizer-hints.md:140]:
```suggestion
EXPLAIN SELECT * FROM t WHERE EXISTS (SELECT 1 FROM t1 WHERE t1.a = t.a);
```

**Review by Oreoxmt — APPROVED** [review]
Rest LGTM

**Pull request merged**